cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -2433,6 +2441,116 @@ pub(crate) fn emit(
+                types::I64 => (LegacyPrefixes::_F0, RexFlags::set_w(), 0x0FB1),
+                _ => unreachable!(),
+            };
+            emit_std_reg_mem(sink, prefix, opcodes, 2, *src, &dst.finalize(state), rex);
+        }
+
+        Inst::AtomicRMW { ty, op, srcloc } => {
+            /* Emit this:
+
+                  mov{b,w,l,q}           (%r9), %rax  // rax = old value
+                 again:
+                  movq                   %rax, %r11   // rax = old value, r11 = old value
+                  `op`q                  %r10, %r11   // rax = old value, r11 = new value
+                  lock cmpxchg{b,w,l,q}  %r11, (%r9)  // try to store new value
+                  jnz again // If this is taken, rax will have a "revised" old value
+
+               Operand conventions:
+                  IN:  %r9 (addr), %r10 (2nd arg for `op`)
+                  OUT: %rax (old value), %r11 (trashed), %rflags (trashed)
+
+               In the case where the operation is 'xchg', the second insn is instead
+                 movq                    %r10, %r11
+               so that we simply write in the destination, the "2nd arg for `op`".
+            */
+            let rax = regs::rax();
+            let r9 = regs::r9();
+            let r10 = regs::r10();
+            let r11 = regs::r11();
+            let rax_w = Writable::from_reg(rax);
+            let r11_w = Writable::from_reg(r11);
+            let amode = Amode::imm_reg(0, r9);
+            let amode_rm = RegMem::mem(amode.clone());
+            let again_label = sink.get_label();
+
+            // mov{b,w,l,q} (%r9), %rax
+            // No need to call `add_trap` here, since the `i1` emit will do that.
+            let i1 = match *ty {
+                types::I8 => Inst::movzx_rm_r(ExtMode::BQ, amode_rm, rax_w, *srcloc),
+                types::I16 => Inst::movzx_rm_r(ExtMode::WQ, amode_rm, rax_w, *srcloc),
+                types::I32 => Inst::movzx_rm_r(ExtMode::LQ, amode_rm, rax_w, *srcloc),
+                types::I64 => Inst::mov64_rm_r(amode_rm, rax_w, *srcloc),
+                _ => unreachable!(),
+            };
+            i1.emit(sink, flags, state);
+
+            // again:
+            sink.bind_label(again_label);
+
+            // movq %rax, %r11
+            let i2 = Inst::mov_r_r(true, rax, r11_w);
+            i2.emit(sink, flags, state);
+
+            // opq %r10, %r11
+            let r10_rmi = RegMemImm::reg(r10);
+            let i3 = match op {
+                AtomicRMWOp::Add => Inst::alu_rmi_r(true, AluRmiROpcode::Add, r10_rmi, r11_w),
+                AtomicRMWOp::Sub => Inst::alu_rmi_r(true, AluRmiROpcode::Sub, r10_rmi, r11_w),
+                AtomicRMWOp::And => Inst::alu_rmi_r(true, AluRmiROpcode::And, r10_rmi, r11_w),
+                AtomicRMWOp::Or => Inst::alu_rmi_r(true, AluRmiROpcode::Or, r10_rmi, r11_w),
+                AtomicRMWOp::Xor => Inst::alu_rmi_r(true, AluRmiROpcode::Xor, r10_rmi, r11_w),
```

Comment:
Can you commonize all these match arms together? `=> Inst::alu_rmi_r(true, AluRmiROpcode::from(op), r10_rmi, r11_w)` to avoid code duplication

Reply:
Done.  Although the result is actually longer:
```
            let i3 = if op ==  AtomicRMWOp::Xchg {
                AtomicRMWOp::Xchg => Inst::mov_r_r(true, r10, r11_w),
            } else {
                let alu_op = match op {
                    AtomicRMWOp::Add => AluRmiROpcode::Add,
                    AtomicRMWOp::Sub => AluRmiROpcode::Sub,
                    AtomicRMWOp::And => AluRmiROpcode::And,
                    AtomicRMWOp::Or => AluRmiROpcode::Or,
                    AtomicRMWOp::Xor => AluRmiROpcode::Xor,
                    AtomicRMWOp::Xchg => unreachable!(),
                }
                Inst::alu_rmi_r(true, alu_op, r10_rmi, r11_w)
            };
```
I tend to assume that rustc/LLVM will do tail-merging and hence cause all the duplication to disappear in the final machine code.  I don't know that that's true, though.